### DISCUSSION TITLE
Integrating tailwind css in theia using React

### DISCUSSION THREAD

**Discussion by ayush-shta** [eclipse-theia/theia#13455] [General]
I'm using react, and I could not get tailwind CSS to work with Theia. Can anyone please help me with this?

**Comment by JonasHelming** [comment]
Could you describe the issue more in detail?

**Reply by ayush-shta** [reply]
I tried installing tailwind css using this [link](https://tailwindcss.com/docs/installation), on our thiea project that uses React for frontend UI.

I tried the steps mentioned in both the [Tailwind CLI](https://tailwindcss.com/docs/installation) and [Using PostCSS](https://tailwindcss.com/docs/installation/using-postcss). But I could not figure out how to integrate within our theia app. 

I think the main issue is that I could not figure out where to how to do the last step mentioned in the docs ("Add your compiled CSS file to the `<head>`  "), since I cannot find where will be our equivalent index.html file is for theia.

I could be thinking this in a complete wrong way. Please let me know what steps should I take for this.

**Reply by sdirix** [reply]
Hi @ayush-shta,

The `index.html` is fully generated during `theia build`. Of course you could modify it afterwards, it will be located in `lib/frontend/index.html`. However it will be overwritten with each build.

However you should be able to include the CSS file progammatically. I think just importing it in some source file of yours, e.g. `import './generated.css';` should be sufficient. It will then be processed by webpack and be automatically included. You just need to make sure that the path of the generated CSS file matches the deployment. You could for example handle this with webpack customization, adding a copy step.

**Reply by ayush-shta** [reply]
Thanks @sdirix for the suggestion. 

However, I'm still a newbie on this. I did not see the generated CSS being generated, and I'm still not sure where or how should I make the webpack customization.

Could you please let me know where we should be looking into?

**Comment by kachurun** [comment]
Its pretty easy man,
First of all you need to drop off tsc from your plugin build process and replace it with vite:

vite.config.mjs
```
import tailwindcss from '@tailwindcss/vite';
import react from '@vitejs/plugin-react';
import { resolve } from 'path';
import { defineConfig } from 'vite';
import { libInjectCss } from 'vite-plugin-lib-inject-css';

export default defineConfig({
    plugins: [
        react(),
        tailwindcss(),
        libInjectCss(),
    ],
    build: {
        lib: {
            entry: resolve(__dirname, 'src/index.ts'),
            formats: ['cjs'],
            fileName: 'index',
        },
        outDir: 'lib',
        emptyOutDir: true,
        cssCodeSplit: true,
        sourcemap: true,
        target: 'ES2022',
        rollupOptions: {
            external: (id) => {
                // Aliases is not external
                if (id.startsWith('@/')) {
                    return false;
                }

                return !id.startsWith('.') && !id.startsWith('/') && !id.includes('\0');
            },
            output: {
                preserveModules: true,
                preserveModulesRoot: 'src',
                entryFileNames: '[name].js',
                chunkFileNames: '[name].js',
                format: 'cjs',
                exports: 'named',
            },
        },
    },
    resolve: {
        alias: [{
            find: /^@\/(.*)/,
            replacement: `${ resolve(__dirname, 'src') }/$1`,
        }],
    },
});

```

It also fix some issues like necessary to use crazy paths to css files, like '../../src/styles/index.css'

Then import your css file in your module and import tailwind in this css file. Easy peasy.

**Reply by kachurun** [reply]
I also recommend replace lerna with turborepo to run everything smoothly. But maybe with lerna everything will work too
